### PR TITLE
Allow multiple instances of rook

### LINKED ISSUE #444: Allow multiple instances of Rook to start
There is a restriction in the operator that multiple instances of Rook cannot be created. This is awkward because TPRs allow multiple instances to be created regardless of whether Rook wants to support it.

Rook can support multiple instances, we just have to be careful about which instance owns which devices. For now we will only allow one instance to use all devices. In the future with #397.

### PR DESCRIPTION
**Pull request by travisn**
Multiple instances of rook can be supported. The only restriction is that storage devices cannot be used by multiple instances or else they will stomp on each other. Thus, the operator for now will only allow one instance to start with all devices. All others will start as directories.

In the future we could allow the devices to be partitioned by instance.

Fixes #444.

### CODE REVIEW

**Comment by kokhang** [comment]
This is kinda scary to include on the eve of our 0.3 release considering that we havent had that much testing on this. To test this, we would have to create multiple rook clusters in different namespaces and then test each cluster with its own Kube app and make sure volume claims and attachment works for every clusters.

Although this is a good feature, i feel we can hold on this for 0.4. Thoughts?

File: pkg/operator/cluster.go [hunk 1]
```diff
@@ -133,28 +129,29 @@ func (c *Cluster) createClientAccess(clusterInfo *cephmon.ClusterInfo) error {
 	}
 
 	// store the secret for the rbd user in the default namespace
```
Comment (jbw976) [inline pkg/operator/cluster.go]: remind me again why the rbd user secret gets put in the default namespace?

File: pkg/operator/operator.go [hunk 2]
```diff
@@ -172,22 +164,43 @@ func (o *Operator) watchTPR(watchVersion string) error {
 
 }
 
+func (o *Operator) startCluster(cluster *Cluster) {
+	if o.devicesInUse && cluster.Spec.UseAllDevices {
```
Comment (jbw976) [inline pkg/operator/operator.go]: how does this devicesInUse flag get set if the operator gets restarted?

File: pkg/operator/operator.go [hunk 3]
```diff
@@ -172,22 +164,43 @@ func (o *Operator) watchTPR(watchVersion string) error {
...
 		ns := c.Spec.Namespace
-		newCluster := newCluster(ns, c.Spec.Version, c.Spec.UseAllDevices, o.factory, o.clientset)
+		existingCluster := newCluster(c.Spec, o.factory, o.clientset)
 		o.stopChMap[ns] = stopCh
-		o.clusters[ns] = newCluster
+		o.clusters[ns] = existingCluster
 		o.clusterRVs[ns] = c.Metadata.ResourceVersion
+
+		logger.Infof("resuming cluster %s in namespace %s", c.Metadata.Name, ns)
+		o.startCluster(existingCluster)
```
Comment (jbw976) [inline pkg/operator/operator.go]: should this be done in a goroutine?

File: pkg/operator/cluster.go (hunk 1 above)
Comment (kokhang) [inline pkg/operator/cluster.go]: This is done to make it easier for users to get a quick start.  The RBD volume plugin expects a secret on the same namespace as the app with a valid ceph token.

**Comment by travisn** [comment]
@kokhang, it's fine to delay this until after 0.3 release, although this early in the release cycle it doesn't seem like we're introducing risk

**Comment by kokhang** [comment]
I was just concerned with the fact that we are planning on releasing today and dont have enough testing around this yet. But if @jbw976 and you are okai with this change, then im on board too :)

File: pkg/operator/operator.go (hunk 2 above)
Comment (travisn) [inline pkg/operator/operator.go]: at restart, devicesInUse under the covers actually will only be applied on the first start of the cluster. The OSD daemonset gets the original value and it won't be updated through the lifetime of the daemonset.

**Comment by jbw976** [comment]
I’m OK with it if it makes the release more complete and it’s not a mainline scenario (e.g., not covered by the readme), with the understanding we collectively seem to have that future milestones need to be more locked down before release.

File: pkg/operator/operator.go (hunk 3 above)
Comment (travisn) [inline pkg/operator/operator.go]: the goroutine is started inside the startCluster() method

**Pull request merged**